### DISCUSSION TITLE
Making own/new NER model from the scratch

### DISCUSSION THREAD

**Discussion by JorGit91** [explosion/spaCy#12570] [Help: Model Advice]
Hi, I am new to Spacy and I'd like to know several things:

1) Can i make my own (custom) NER model in English (or any other language)  from scratch (i.e. own entities, training data, etc)? or do I have to use models that have already been created?

2) If I make my new ner entity (say chemical/address detector) in any given language and would like to use it in addition to an existing model :  will it detect chemical/address and Person, dates, Org at the same time ? or will I face the catastrophical memory loss problem?

3) My  NER model needs to be able to identify chemicals/addresses :  Do i have to train my data within sentences ? ,i.e. : "Alcohol is an organic compound" [0:6 'Chemical'] or "I live in Elm Street" [10:19 'Street'] or can I just train it as Alcohol [0:6 'Chemical'] and Elm Street" [0:9 'Street']?

Thanks in advance

**Comment by rmitsch** [comment]
Hi @JorGit91!

1. Yes, you can train your own models from scratch. One potential starting point for you is this [NER tutorial](https://github.com/explosion/projects/tree/v3/tutorials/ner_drugs) (and, of course, the [docs](https://spacy.io/api/entityrecognizer)).

2. It is possible to base your new model on an existing one (see [sourcing](https://spacy.io/usage/processing-pipelines#sourced-components)) and continue training with only examples of your new label. In theory the model might be able to recognize all the old and your new labels. In practice it often does lead to catastrophic forgetting, hence we recommend training your model from scratch. If you really want to extend a pretrained NER model by a new label, you should look into [rehearsing](https://spacy.io/api/pipe#rehearse) (again, we don't recommend this).
Alternatively you can also use two NER components in your pipeline - the pretrained one and one that identifies your new entities (e. g. chemicals). See [here](https://github.com/explosion/projects/tree/v3/tutorials/ner_double) for an example. 

3. You should train your model with full sentences/docs, as this is how the data will look like at inference time.

**Reply by JorGit91** [reply]
Hi @rmitsch!

Thanks for the answer! 

I just want to make it clear that I want my model(s) to identify entities without a context/sentence in columns of data (csv, excel, json) : "John Obama", "Alcohol", " Developer" should be labeled as "People", "Chemical", "Job title". Will spacy still work or does the data have to be in a sentence/context in order to make a prediction(classification) ?

1. Is there any maximum/ amount entities that Spacy can optimally handle (i.e. perform NER)? I want to have a model up to 50 entities

2. Can i make an existing model to be more accurate? 

3.Could you please provide me some links on how to train my custom model/ data that was used in order to train existing models?

**Reply by rmitsch** [reply]
The NER component might still work to some degree, but it's not the right tool for this job. The idea behind it is to leverage the sentence context to identify which parts of a text are entities and which entity label they should have.  [Textcat](https://spacy.io/api/textcategorizer) might be a better choice here, but even with Textcat this is problematic, as it's training a model based on a set of word embeddings and some of your target values might not appear in this set.

E. g. "John Obama" might not be part of the vocabulary, in which case you won't be able to fetch embeddings for this string. You could of course combine the embeddings for "John" and "Obama", which have a higher chance of being part of the vocabulary individually. You probably see where this is going though - the more uncommon terms your data contains, the more likely it is you can't retrieve embeddings for them, which in turn means that Textcat won't work properly as it relies on embeddings.

Regarding your questions:
1. There is no maximum number of entities. Consider that the higher the number of entities, the more training data you need.
2. You can continue training a model with more data using [sourcing](https://spacy.io/usage/processing-pipelines#sourced-components). Note that this is only recommended if you continue training with the same labels as the existing models.
3. Check out the [spaCy docs](https://spacy.io/usage/training) on training your own model. There are also example projects on training [NER](https://github.com/explosion/projects/tree/v3/tutorials/nel_emerson) and [Textcat](https://github.com/explosion/projects/tree/v3/tutorials/textcat_goemotions) models.

**Reply by JorGit91** [reply]
What do you mean exactly with "might still work" or "it's not the right tool for this job." or "the more uncommon terms your data contains"? That the data (without context/sentence) will likely become more ambiguous for example? or that the NER component might not classify unseen data (that is not in context/sentence) at all? So far I have been able to tag 2400 Person names on that file, but , yes I did get names like Mercedes Colton or April Smith labeled as ORG and DATE respectively. Is this what you mean regarding context? 

If what you meant is that ambiguity might be my biggest issue here, then I dont think so. 
By the way all the entities I am aiming to classify are quite different from one another: Name, address, date, job title, gender, number, etc

**Reply by rmitsch** [reply]
"might still work" means NER training will terminate successfully and you will be able to use your model to infer entities from your data, but it's questionable how accurate the results will be (probably not very much).

"it's not the right tool for this job." means that your use case does is not what's commonly understood as NER, hence also not what spaCy's NER component was built for. The usual assumption is that NER is run on a natural language corpora corpus, i. e. full sentences with context surrounding potential entity mentions. If there's no context, NER models won't be very helpful, as they are designed to take advantage of this context.

"the more uncommon terms your data contains" means that you'll find common terms in spaCy's models' vocabulary, correspondingly you'll also be able to retrieve word embeddings for these terms. If a word is not in the model's vocabulary however, there is also no embedding for this term. Since both NER and Textcat model work with embeddings, they'll lack this source of information. If there is neither context nor a vector for the actual mention, they won't be able to learn and make informed decisions.

> By the way all the entities I am aiming to classify are quite different from one another: Name, address, date, job title, gender, number, etc

The issue with your data is not that the categories aren't distinguishable enough, it's that you don't provide a lot of information to the model - there is no context and especially for names (or e. g. organisation names) there might not always be a word vector available.

**Comment by JorGit91** [comment]
Thanks for the response.

Last question : Would you be able to tell/guess which tool would be appropriate for this job?

**Reply by rmitsch** [reply]
That depends on your data. If it's exclusively or mostly terms that are in the model vocabulary / you have embeddings for, Textcat could work well enough. Considering your use case though I would strongly recommend collecting contextual data (i. e. whole sentences) - then you could do this with NER or [Spancat](https://spacy.io/api/spancategorizer).

**Reply by JorGit91** [reply]
You mean that i should get contextual data (example "Hi, my name is John Smith, "Alcohol is a chemical compound ) and then then use my own NER fro mthe scratch on that? Now im more confused (lol) : isnt NER supposed to work (classify) only on entities in context? Remember that my target data is not in context: It is only : John Smith, Alcohol, Data scientist

**Reply by rmitsch** [reply]
> You mean that i should get contextual data (example "Hi, my name is John Smith, "Alcohol is a chemical compound ) and then then use my own NER fro mthe scratch on that?

Yes.

> Remember that my target data is not in context: It is only : John Smith, Alcohol, Data scientist

Same recommendation here: same as you should _train_ on your data with context, you should _infer_ on it with context. Imagine e. g. "John Smith" - likely it's a name, but it can also be a [company](https://www.johnsmith.co.uk/). Without context it's impossible for your model to make this distinction. If your data however is "John Smith visited me today" or "I picked up a book from the John Smith store downtown", your NER model can pick up on that.

**Reply by JorGit91** [reply]
Good . But how should i put the target data then for classification if its only without a sentence thou? So far I have put the columns data into a text: "John Smith. me@mail.com, 11-09-1991" and its been quite accurate

Now i see your point: But that confirms what i asked you lines above: So far I havent seen that ambiguity is a problem..but it might be with other categories (example : street names) 

Anyways my whole point with Spacy/Ner is that it can be used for my problem or not at all?

**Reply by JorGit91** [reply]
Bye the way, putting the data in the way above: "John Smith. [me@mail.com](mailto:me@mail.com), 11-09-1991" wouldn't doing so give the target data a context, i.e. would this be a valid way of putting it into context?

**Reply by rmitsch** [reply]
> But how should i put the target data then for classification if its only without a sentence thou?

I recommend to collect the sentences surrounding these entities, if possible.

> Anyways my whole point with Spacy/Ner is that it can be used for my problem or not at all?

Yes, it can be used. The accuracy of the model would benefit from access to contextual data.

**Reply by rmitsch** [reply]
> Bye the way, putting the data in the way above: "John Smith. [me@mail.com](mailto:me@mail.com), 11-09-1991" wouldn't doing so give the target data a context, i.e. would this be a valid way of putting it into context?

It's possible that this is helpful, as long as the context for your other labels looks different. I recommend using full sentences mentioning your entities though.

**Reply by JorGit91** [reply]
Nice! Would you give an example of how you'd put it if your target data are only columns containing those entities? And yes, the other entities look different (thus almost no ambiguity)

**Reply by rmitsch** [reply]
> Would you give an example of how you'd put it if your target data are only columns containing those entities?

Sorry, I don't understand your question.

**Reply by JorGit91** [reply]
By the way : If i tell SPacys NEr model to "switch off" ORG when performing the classification: Will the risk for miss classifications be less when inputting the target data? Example : April Smith (without context)  could be easily classified as either DATE or PERSON ..but switching DATE off would help avoiding this miss classification?

**Reply by JorGit91** [reply]
Would you give an example of how you'd put it if your target data are only columns containing those entities? Here I meant how youd put John Smith, email or date into a context when wanting to perform NER

**Reply by rmitsch** [reply]
> If i tell SPacys NEr model to "switch off" ORG when performing the classification: Will the risk for miss classifications be less when inputting the target data?

This is not possible without retraining your model.

> Would you give an example of how you'd put it if your target data are only columns containing those entities?

There is no one correct way to do this, I suggest experimenting. I want to highlight again that you're applying NLP methods (i. e. NER) in a situation that those methods weren't designed for, because you _don't have_ natural text (as in full sentences). You may want to look into other methods as sell, such as rule-based matching or looking up word lists.

**Reply by JorGit91** [reply]
ok. One last questions : Do you think making a permanent template of natural text would work? I am thinking something like "Hi I am (data to be identified with ner) and my email is (data to be identified by NER)" or does this text have to change continuously?

**Reply by rmitsch** [reply]
Yes, that would probably help. If you already know in advance which bits of your data are person names or emails or whatever, why would you want to run NER (or any other approach) on it though?

**Reply by JorGit91** [reply]
If you already know in advance which bits of your data are person names or emails or whatever, why would you want to run NER (or any other approach) on it though?

Becuase I need to make sure that the data under some header (say chemicals) are what they suppose to be. Anyways I am thinking of a more slightly "unstructred" text that isnt that indicative but still has some context. 

I sill keep getting **very good results** with name tagging using this format: "John Smith. myaccount@email.com. italian. 3.4.2023.  But of course, names like April Roberts gets classified as date or ORG. Could you explain a little further what you meant  with "theres might be no vector to represent this string" ? Could this be fixed (i.e. probability theres no vector for a string) by just adding vectors to my model = the more vectors the least likely a given string wont have any?
I will give you an example of my template text when i finish it this afternoon

**Reply by JorGit91** [reply]
Template text: Chen Li. [acbuchli@gmail.com](mailto:acbuchli@gmail.com). EDUCATION, Northeastern University, Seattle, WA. WORK EXPERIENCES. At Home Store LLC. Data Scientist. January 2019, June 2020.

Something like the example above gets classified pretty good

**Reply by rmitsch** [reply]
> Something like the example above gets classified pretty good

If it's working for you, great! In general the train data should resemble the test data as closely as possible.

**Reply by JorGit91** [reply]
Great. What about the question about vectors above?

**Reply by JorGit91** [reply]
..And can i see how existing models where trained? If so, could you pint out where?

**Reply by adrianeboyd** [reply]
The underlying issue is that you need context to disambiguate words like "April" as DATE vs. PERSON. People would need context to disambiguate this (like with the "John Smith" example) and NLP models also need context to disambiguate this.

For the default spacy models like NER, this context comes from the surrounding text. Word vectors might help a little in cases like "April Roberts", but the static word vector for "April" will be the same for all occurrences of "April", so what the NER model mainly needs is more context.

If you don't have context, then spacy's NER model is probably not the right choice and a rule-based approach with lists or patterns for known entity types might be a better approach. You can still use a spacy pipeline, but you might consider an entity ruler or span ruler instead of NER.

**Reply by JorGit91** [reply]
Hi @adrianeboyd !

If the main issue is ambiguity : wouldn't a costumed model (with only names in it for example) help with the issue if i only want to label/detect names? and wouldn't a sentence like "Hi, my name is April" would help when classifying v/s dates? or do we need a whole text to label a single word?

The main thing with my task is to be able to tell that a given string is a Name, so rule-based methods would be lengthy and not beneficial at all in the long run.

As pointed out above: I have put my data like "John Smith. [myaccount@email.com](mailto:myaccount@email.com). italian. 3.4.2023" and it has yielded very good results (full names get classified almost 96-98% of all of them) , except for those that are ambiguous: like DATE or ORG.

My whole point here with NER is that if it is able to detect/tagg names at all : i.e. if it has learned (i.e.  that NER isn't  a "list" based method ) to recognize peoples names and not that if a new name comes  it wont be able to tell whether it is a name or not ? (of course some exotic names not but this is can be fixed by updating the method with new data).
